### PR TITLE
#184581179 Two Factor Authentication for Sellers

### PR DESCRIPTION
**Pull request by wilbrord2**
#### What does this PR do?

This PR verify if user is a seller and he/she can login as a seller

#### Description of Task to be completed?

- [x] A seller will be successfully login with email and password

- [x] A seller should receive email containing random number for 2FA to verify if he/she is owner of stall.

- [x] Document feature using swagger.

#### How should this be manually tested?
- Clone this repo `git clone https://github.com/atlp-rwanda/e-comm-team-amigos-bn`
- Switch your working directory `cd e-comm-team-amigos-bn`
- Create an` .env` file and update it with `.env-example`
- Install all the dependencies by running` npm install`
- Then run the project by typing `npm dev` to start server
- Make sure that postman is installed, open postman or any other API testing tool of your choice
- choose route to be `POST`
- Navigate `localhost:3000/user/login`
- Make sure you created account, enter your username and password
- if you are a seller you will receive email containing `OTP (one-time-password) `
- you can use it to successfully login

#### Endpoints to be tested:

- `POST user/login`

#### Any background context you want to provide?

- PostgreSQL installed in your local machine
- Node JS installed in your local machine

#### What are the relevant pivotal tracker/Trello stories?

Story ID: [#184581179](https://www.pivotaltracker.com/story/show/184581179)

#### Screenshot of  `swagger documentation`  ,  `test`   and  `email format`  to be sent

Screenshot (134)

Screenshot (133)

email

### CODE REVIEW

File: src/tests/auth.test.js [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    await models.sequelize.sync();
+    const result = await createOTP(user);
+    expect(result).to.have.property('status', 'OK');
+    expect(result).to.have.property('otp').that.is.a('string');
+    expect(result).to.have.property('email').that.is.a('string');
+
+    // Check that the user's otpcode and otpcodeexpiration have been updated in the database
+    await models.sequelize.sync();
+    const updatedUser = await models.User.findByPk(user.id);
+    // const timestamp = updatedUser.otpcodeexpiration;
```
Comment (nshutijonathan) [inline src/tests/auth.test.js]: Don't push commented out codes

File: src/tests/auth.test.js [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    expect(updatedUser).to.have.property('otpcode').that.is.a('string');
+    expect(updatedUser).to.have.property('otpcodeexpiration').that.is.a('string');
+    // sinon.assert.calledOnce(sendMailStub);
+  });
+
+  // it('should send an email', async () => {
+  //   const transporterSpy = sinon.spy(nodemailer, 'createTransport');
+  //   const mailOptions = {
+  //     from: 'sender@example.com',
+  //     to: user.email,
```
Comment (nshutijonathan) [inline src/tests/auth.test.js]: commented out codes should not get pushed

**Review by nshutijonathan — CHANGES_REQUESTED** [review]

**Review by GasanaJr — CHANGES_REQUESTED** [review]
The feature works as expected, but the circleci tests must still be included. Integrate the tests, and we will be good to go. Also update the body of your PR to include screenshots for swagger and running tests.

**Review by 1MuhireDavid — APPROVED** [review]

**Review by Rwibutso — APPROVED** [review]

**Review by GasanaJr — CHANGES_REQUESTED** [review]
Solve conflicts you have in develop

**Review by AbdulKhaliq59 — APPROVED** [review]

**Review by AbdulKhaliq59 — CHANGES_REQUESTED** [review]
Resolve conflict

**Review by Emile-SHUMBUSHO — APPROVED** [review]

**Review by GasanaJr — APPROVED** [review]

**Review by AbdulKhaliq59 — APPROVED** [review]

File: src/database/migrations/20230312135251-Users.js [hunk 3]
```diff
@@ -60,8 +68,20 @@ module.exports = {
         type: Sequelize.DATE
       }
     });
+    // await queryInterface.addColumn('Users', ' otpcode', {
```
Comment (nshutijonathan) [inline src/database/migrations/20230312135251-Users.js]: remove commented out codes here

File: src/database/migrations/20230312135251-Users.js [hunk 4]
```diff
@@ -60,8 +68,20 @@ module.exports = {
...
+    //   type: Sequelize.STRING,
+    //   allowNull: true
+    // });
+
+    // await queryInterface.addColumn('Users', 'otpcodeexpiration', {
+    //   type: Sequelize.STRING,
```
Comment (nshutijonathan) [inline src/database/migrations/20230312135251-Users.js]: remove commented out codes here

File: src/database/migrations/20230312135251-Users.js [hunk 5]
```diff
@@ -60,8 +68,20 @@ module.exports = {
...
+    //   allowNull: true
+    // });
   },
-  async down(queryInterface, Sequelize) {
+
+  async down(queryInterface) {
     await queryInterface.dropTable('Users');
+    // await queryInterface.removeColumn('Users', 'otpcode');
+    // await queryInterface.removeColumn('Users', 'otpcodeexpiration');
```
Comment (nshutijonathan) [inline src/database/migrations/20230312135251-Users.js]: commented out codes should not be pushed

**Review by nshutijonathan — CHANGES_REQUESTED** [review]
Remember to rebase so that the conflicts are gone and then we are good to go

**Review by nshutijonathan — APPROVED** [review]
LGTM

**Pull request merged**